clusterloader2/pkg/measurement/util/object_store.go

```diff
@@ -98,13 +99,53 @@ func NewPodStore(c clientset.Interface, selector *ObjectSelector) (*PodStore, er
 }
 
 // List returns list of pods (that satisfy conditions provided to NewPodStore).
-func (s *PodStore) List() []*v1.Pod {
+func (s *PodStore) List() ([]*v1.Pod, error) {
 	objects := s.Store.List()
 	pods := make([]*v1.Pod, 0, len(objects))
 	for _, o := range objects {
 		pods = append(pods, o.(*v1.Pod))
 	}
-	return pods
+	return pods, nil
+}
+
+type controlledPodsIndexer interface {
+	PodsControlledBy(obj interface{}) ([]*v1.Pod, error)
+}
+
+// OwnerReferenceBasedPodStore returns pods with ownerReference set to the owner
+// (with exception for Deployment, see ControlledPodsIndexer implementation) and matching given labelSelector.
+type OwnerReferenceBasedPodStore struct {
+	controlledPodsIndexer controlledPodsIndexer
+	owner                 interface{}
+	labelSelector         labels.Selector
+}
+
+func NewOwnerReferenceBasedPodStore(controlledPodsIndexer controlledPodsIndexer, owner interface{}, labelSelector labels.Selector) *OwnerReferenceBasedPodStore {
+	return &OwnerReferenceBasedPodStore{
+		controlledPodsIndexer: controlledPodsIndexer,
+		owner:                 owner,
+		labelSelector:         labelSelector,
+	}
+}
+
+// List returns controlled pods.
+func (s *OwnerReferenceBasedPodStore) List() ([]*v1.Pod, error) {
+	pods, err := s.controlledPodsIndexer.PodsControlledBy(s.owner)
+	if err != nil {
+		return nil, err
+	}
+
+	// Using ownerReference as proxy for pods controlled by a given object should be pretty accurate,
+	// but technically the API is defined on labelSelectors.
+	// Let's check them to work properly in uncommon cases (like e.g. change of label selector in deployment).
+	// TODO(mborsz): Is it really needed?
```

Comment:
I don't think it's really needed - it can only speed up things a bit.

Technically what every controller is doing is when the pod changes (and it was previously controller by it) it will orphan it (by removing its ownerRef). So checking for it is only an optimization.

If you would like to be 100% accurate you would need to go over all other pods and see if others shouldn't be picked up (which controller didn't yet adopt). But again - it's an optimization as controller will eventually adopt them.

Reply:
I suggest adding a comment along the lines what I wrote above instead and getting rid of this code.